gitpod-io/gitpod
# Add a persisted query client provider

## Description

Adding a persisted query cache provider for react-query in the Dashboard. This will give us some cache benefits between page loads where the app will load with the query cache primed to whatever it last was. A few details on how it works:

* The query client is stored in Indexed-DB, which offers a much higher storage limit than localstorage, and also doesn't require serialization. Browser support for indexed-db is [quite good now](https://caniuse.com/indexeddb).
* The `maxAge` of the persister determines how long a cached client is valid for. This doesn't necessarily mean individual queries cached in it will still be valid, just that when the app loads, and we look for a persisted client, if one is present and newer than `maxAge`, it will be used. Each query in it's cache will be evaluated for staleness/expired based on the query settings (they all default to 5 minutes, but we can/should tune some of those to be longer).
* The `buster` is a simple version string that we can manually increment if we know the shape of cached queries changes in a backwards incompatible way. This will give us a mechanism to force-prune any cached clients after a deploy. I think it will be interesting to see if we can find ways to automate or add tests around the shapes changing to help us know this needs to be bumped as we learn more.

## Related Issue(s)

Fixes #15674

If you're curious, you can pop open the devtools Application tab, and take a look at the data stored in IDB
![image](https://user-images.githubusercontent.com/367275/213754539-c6b5353a-19c0-4361-8744-e8ade4513843.png)


## How to test

* For starters, on the preview env create a team and project and a workspace.
* After navigating to the workspaces and projects page (which primes the query cache), go to the workspaces list and reload the page.
* It's hard to spot, but the list should load w/ data rendered and not have a loading indicator.
* Navigate to the Projects page and it should instantly have data (no loading spinner).

* For a more exhaustive test, open the preview environment in a different browser than what you used for your first test (not just a new tab/profile, but Safari, or Firefox/Brave/Chrome w/e).
* Load the Projects page in this new browser, keep track of what projects are there.
* In the first browser, create a new project.
* Go back to the 2nd browser. You shouldn't see that new Project yet. Reload the page.
* You should see it load w/o the new project (means it loaded from the persisted client).
* The new project should shortly show up after you reloaded (since a new query was kicked off).

### Browser Testing
* I've done some fairly in-depth testing of different browsers to see how it behaves. All of the following have worked fine and the client gets persisted/loaded as expected.
  * Chrome/Brave/Arc Mac OS
  * Safar Mac OS
  * Firefox Mac OS
  * Mobile Safari iOS
  * Chrome iOS
  * Edge Windows
  * Chrome Windows
* Private Browsing mode in some browsers disabled indexed-db usage. In this case, we degrade gracefully and just don't store/retrieve the client from idb.

## Release Notes

```release-note
NONE
```

## Documentation


## Build Options:

      Experimental feature to run the build with GitHub Actions (and not in Werft).
      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
